src/Common/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -57,24 +57,29 @@ public static int ListHashCode<T>(this IEnumerable<T> list)
         }
 
         [Pure]
-        public static bool ListEquals<T>(this ICollection<T> first, ICollection<T> second)
+        public static bool ListEquals<T>(this ReadOnlyCollection<T> first, ReadOnlyCollection<T> second)
         {
-            if (first.Count != second.Count)
+            if (first == second)
+            {
+                return true;
+            }
```

Comment:
Before your added tests, does the return true you added here show as being hit in code coverage?